### PR TITLE
Set up golden tests

### PR DESCRIPTION
**Pull request by thealmarty**
- set up golden tests using `Tasty.Silver`.
- add Id contract golden test.

### CODE REVIEW

File: test/examples/Id-Strings.ju.golden
Comment (mariari) [inline test/examples/Id-Strings.ju.golden:1]: Can we run a formatter on this, so it's not one big line?

Will that mess up how the golden tests work?

**Comment by thealmarty** [comment]
> tACK, though can we reduce the output when running `make test`? I don't think we need to print the parse results when the test passes, only when it fails.

I'm thinking we can predominately use the golden test but when a contract doesn't pass we can use the stdout to debug because it prints out more output. So the stdout out would only have one file (`test.ju`).

File: test/examples/Id-Strings.ju.golden
Comment (thealmarty) [reply test/examples/Id-Strings.ju.golden:1]: I will do more testings and see if there is a way. I fear that if we change the golden file in anyway the tests would fail. But I'll see what I can do.

Comment (mariari) [reply test/examples/Id-Strings.ju.golden:1]: If it just parses it as a haskell string there is a way to trick it, probably the same thing I've done in the FrontendTests (adding \ to the end and start to signal haskell where to begin again

**Comment by cwgoes** [comment]
We should probably only be committing one of the files (whatever one Golden keeps)?

**Comment by thealmarty** [comment]
In this latest commit I use `readMaybe` instead of `read`, which gives a different error. But it's really just the same as the last commit's: `read` has an exception `Prelude.read: no parse`. 

I have verified that at the `readFileMaybe` step it *did* return a `Just ByteString`, so that step isn't the problem. Working on debugging the rest of the pipeline towards `read`.

**Comment by thealmarty** [comment]
Verified in this commit that using derived `Read` and `Show` instances for `Usage` and `Symbol` doesn't fix the error. Probing further.

**Comment by thealmarty** [comment]
Debugging in progress - will be testing more `TopLevel`s and figure out which one's instance may be causing the problem.

**Comment by mariari** [comment]
> Debugging in progress - will be testing more `TopLevel`s and figure out which one's instance may be causing the problem.

you can try parsing expressions, then invoking the read instances back on it modulo the quotes.

**Comment by thealmarty** [comment]
- Custom instances for `Usage` and `Symbol` replaced by derived instances.
- golden test now compares ASTs directly.
- pretty-show function to be finished in another PR - right now it doesn't make anything prettier.
- other frontend tests to be fixed in another PR. They will be changed to comparing the ASTs instead of their strings.

File: test/examples/Id-Strings.ju.golden
Comment (mariari) [inline test/examples/Id-Strings.ju.golden:1]: Can you run the formatter on all these files

File: test/examples/Token.ju.golden
Comment (mariari) [inline test/examples/Token.ju.golden:1]: Ditto

File: test/examples/Addition.ju.golden
Comment (mariari) [inline test/examples/Addition.ju.golden:1]: Ditto

File: test/Main.hs
Comment (mariari) [inline test/Main.hs:55]: We need all our tests back

File: test/Frontend.hs
Comment (mariari) [inline test/Frontend.hs:42]: we need all these back as well

File: src/Juvix/Library.hs
Comment (mariari) [inline src/Juvix/Library.hs:82]: This seems like an unused import?

File: src/Juvix/Core/Usage.hs
Comment (mariari) [inline src/Juvix/Core/Usage.hs:16]: Same here

File: test/examples/Id-Strings.ju.golden
Comment (thealmarty) [reply test/examples/Id-Strings.ju.golden:1]: As mentioned in the comment I'll do the pretty show in another PR.

File: test/Main.hs
Comment (thealmarty) [reply test/Main.hs:55]: I will put them back once I update the test to compare the ASTs instead of strings.

File: src/Juvix/Library.hs
Comment (thealmarty) [reply src/Juvix/Library.hs:82]: Yes, will remove them.

File: test/examples/Id-Strings.ju.golden
Comment (mariari) [reply test/examples/Id-Strings.ju.golden:1]: Well would still be nice since this pr uses the read instance, and since the originals were split up per line for readability reasons, we can do even better formatting with the read instance here!

**Comment by cwgoes** [comment]
Looks like `make test` is still failing?

**Comment by thealmarty** [comment]
> Looks like `make test` is still failing?

Yeah still debugging. It can't find the golden files now. :( Should get it solved by end of today I hope. Another batch of tests will be added back in too.

**Comment by thealmarty** [comment]
Tests updated and cleaned up. R4R. Sorry about the large PR and diffs - ignore changes that are not related to tests.

File: test/examples/Id-Strings.ju.golden
Comment (thealmarty) [reply test/examples/Id-Strings.ju.golden]: I'll format the golden files in another PR. Just really want this to merge first. All tests pass now.

File: test/Frontend/Parser.hs
Comment (mariari) [inline test/Frontend/Parser.hs:167]: Where did this get moved to again?

File: test/Frontend/Parser.hs
Comment (mariari) [inline test/Frontend/Parser.hs:207]: This needs to be newlined... it's 281

File: test/Frontend/Parser.hs
Comment (mariari) [inline test/Frontend/Parser.hs:219]: Same here

File: test/Frontend/Parser.hs
Comment (mariari) [inline test/Frontend/Parser.hs:231]: Same here

File: test/Frontend/Parser.hs
Comment (mariari) [inline test/Frontend/Parser.hs:624]: I'll probably just conver this myself later removing so many of the lines

File: test/examples/Addition.ju.golden
Comment (mariari) [inline test/examples/Addition.ju.golden:1]: This is still 1 line!

File: test/examples/Id-Strings.ju.golden
Comment (mariari) [inline test/examples/Id-Strings.ju.golden:1]: Same here!

File: test/examples/Token.ju.golden
Comment (mariari) [inline test/examples/Token.ju.golden:1]: Same here!

File: test/Main.hs
Comment (mariari) [inline test/Main.hs:20]: Can we change all these back?

I really don't want unqualified imports, it is poor that we have redundant names, but we should change the main test name not the qualified itself

File: test/Frontend/Desugar.hs
Comment (mariari) [inline test/Frontend/Desugar.hs:47]: I'm not sure the desugar test should do this...  I don't see where we desugar

File: test/Frontend/Desugar.hs
Comment (mariari) [inline test/Frontend/Desugar.hs:61]: I don't think we desugar at all here!

File: test/Frontend/Desugar.hs
Comment (mariari) [inline test/Frontend/Desugar.hs:35]: I'd much rather keep the qualified imports and deal with the verbosity by using |>'s

**Review by mariari — CHANGES_REQUESTED** [review]
I will approve when

1. Once the *.golden files have been ran through our formatter
2. the desugar tests actually desugar again!
3. the qualified imports for the main are back

I will handle the formatting for the tests to remove the redundant ()'s via |> after this merges

File: test/Frontend/Parser.hs
Comment (mariari) [reply test/Frontend/Parser.hs:207]: I'll fix this up myself after the pr merges

File: test/Frontend/Parser.hs
Comment (mariari) [reply test/Frontend/Parser.hs:219]: I'll fix this up later as well

File: test/Frontend/Desugar.hs
Comment (cwgoes) [reply test/Frontend/Desugar.hs:47]: We should be calling `desugar` somewhere, right?

File: test/Frontend/Golden.hs
Comment (cwgoes) [inline test/Frontend/Golden.hs:115]: I don't quite understand, what isn't working here / why?

**Review by cwgoes — COMMENTED** [review]
Let's ensure that the tests are still testing the right things, I'm not so concerned about the `.golden` file format as long as it is consistent.

File: test/Frontend/Golden.hs
Comment (thealmarty) [reply test/Frontend/Golden.hs:115]: When a test fails and shows the golden/actual value, it doesn't use this function in the print out. It also doesn't use the next input to update the golden file.

Comment (cwgoes) [reply test/Frontend/Golden.hs:115]: Hmm, that's odd, sounds like either a library bug or we're using the library incorrectly somehow?

File: test/Frontend/Desugar.hs
Comment (thealmarty) [reply test/Frontend/Desugar.hs]: Oops corrected.

File: test/Frontend/Golden.hs
Comment (thealmarty) [reply test/Frontend/Golden.hs:115]: Yeah I don't know if I misunderstood the function or what. There's isn't a lot of documentation about this that I can find.

**Comment by thealmarty** [comment]
- Desugar test actually desugars now. 
- Per meeting more formatting and golden file formats will be fixed by @mariari . 
- Golden test function to be fixed later. If we can't find how to use those "show" and "update golden file" functions, I can change it back to the old test which doesn't take those inputs.

**Review by mariari — APPROVED** [review]

**Comment by cwgoes** [comment]
Alright. We should check why the `test_suite` is still failing though.

**Comment by mariari** [comment]
> Alright. We should check why the `test_suite` is still failing though.

agreed

**Comment by thealmarty** [comment]
> > Alright. We should check why the `test_suite` is still failing though.
> 
> agreed

Seems to have timed out. It succeeded in my local run. I re-ran it, some timed out error below:
image

**Pull request merged**
